### PR TITLE
add recursive locks, cleanup code, tests for cleanup code

### PR DESCRIPTION
**Pull request by earonesty**

### CODE REVIEW

File: src/hilok.cpp [hunk 1]
```diff
@@ -1,84 +1,130 @@
 #include "hilok.hpp"
 #include "psplit.hpp"
 
-bool lock_with_params(std::shared_timed_mutex &mut, bool block, int timeout) {
+bool lock_with_params(HiMutex &mut, bool block, int timeout) {
```
Comment (earonesty) [inline src/hilok.cpp]: change to float/double ... integer not enough for all use cases `std::chrono::duration<double>(timeToSleep)`

File: src/hilok.cpp [hunk 2]
```diff
@@ -1,84 +1,130 @@
...
     }
 
-    return HiHandle(false, refs);
+    return HiHandle(*this, false, refs);
 }
 
+
+void HiLok::erase_safe(HiKeyRef &ref) {
+    std::lock_guard<std::mutex> guard(m_mutex);
+    if (ref.m_mut.use_count() <= 3) {
```
Comment (colatkinson) [inline src/hilok.cpp]: Why 3?

File: src/hilok.hpp [hunk 3]
```diff
@@ -2,34 +2,166 @@
...
+
+    bool try_lock_shared() {
+        if (m_mut.try_lock_shared()) {
+            _add_lock_shared();
+            return true;
+        }
+        return false;
+    }
+
+    bool try_lock_shared_for(int secs) {
```
Comment (colatkinson) [inline src/hilok.hpp]: What's the behavior when secs <= 0?

File: src/hilok.cpp [hunk 4]
```diff
@@ -1,84 +1,130 @@
...
+            }
+
+            refs.emplace_back(mut, key);
+            cur = (void *) mut.get();
         }
     } catch (...) {
-        HiHandle(true, refs).release();
+        auto hh = HiHandle(*this, true, refs);
+        refs.clear();
+        hh.release();
```
Comment (colatkinson) [inline src/hilok.cpp:128]: Is it possible to have `HiHandle`'s destructor automatically release its resources, so that we can take advantage of RAII here?

File: src/hilok.cpp [hunk 5]
```diff
@@ -1,84 +1,130 @@
...
-            cur = (void *) mut;
+            }
+
+            refs.emplace_back(mut, key);
+            cur = (void *) mut.get();
         }
     } catch (...) {
-        HiHandle(true, refs).release();
+        auto hh = HiHandle(*this, true, refs);
+        refs.clear();
```
Comment (colatkinson) [inline src/hilok.cpp]: What's the need for this?

File: src/hilok.cpp [hunk 6]
```diff
@@ -1,84 +1,130 @@
...
     } else {
         mut.lock_shared();
         return true;
     }
 }
 
+void HiHandle::release() {
+    for (auto it = m_refs.begin(); it!= m_refs.end(); ) {
+        auto &kref = *it;
+        ++it;
```
Comment (colatkinson) [inline src/hilok.cpp:49]: Seems like this'll end up not unlocking the first entry in `m_refs`, unless I'm missing something.

File: src/hilok.cpp [hunk 7]
```diff
@@ -1,84 +1,130 @@
...
-            std::shared_timed_mutex *mut;
-            {
-                std::lock_guard<std::mutex> guard(m_mutex);
-                mut = &m_map[key];
-            }
+            std::shared_ptr<HiMutex> mut = _get_mutex(key);
             bool ok = shared_lock_with_params(*mut, block, timeout);
-            if (!ok)
+            if (!ok) {
+                mut.reset();
```
Comment (colatkinson) [inline src/hilok.cpp]: Why is this necessary? Would have thought `mut` would get cleaned up when it goes out of scope on the next line

File: tests/test.cpp [hunk 8]
```diff
@@ -106,6 +181,33 @@ TEST_CASE( "many-threads", "[basic]" ) {
...
+    {
+        threads.emplace_back(std::thread([&] () { nesty_worker(i, h, ctr); } ));
+    }
+
+    for (auto& thread : threads) {
+        thread.join();
+    }
+    CHECK(ctr == pool_size);
+    CHECK(h.size() == 0);
 }
```
Comment (colatkinson) [inline tests/test.cpp:351]: Test that might be interesting: workers all take out write on differing depths (e.g. T1 does `h.write("a/b/c")`, T2 `h.write("a/b")`, T3 `h.write("a/b/c/d")`). They should all still conflict (T1 would block both T2 and T3), and would likely exercise bits of the refcounting code.

File: tests/test.cpp [hunk 9]
```diff
@@ -5,6 +5,8 @@
 
 #include <thread>
 
+#include <iostream>
+
```
Comment (colatkinson) [inline tests/test.cpp:9]: Would be interested in the results of running the tests with TSan -- not necessarily in CI even, just as a one-off. If everything works there we'd have fairly high certainty that there aren't any races.

File: tests/test.cpp [hunk 10]
```diff
@@ -42,8 +44,26 @@ TEST_CASE( "path-split-one", "[basic]" ) {
     REQUIRE(res == expect);
 }
```
Comment (colatkinson) [inline tests/test.cpp:53]: Dumb additional psplit test that might be worthwhile: passing in `""`. AFAICT it should work fine, but easy to break if refactoring.

File: src/hilok.cpp (hunk 2 above)
Comment (earonesty) [reply src/hilok.cpp]: literally no idea, scary really.   tried to get it down to 2 consistently.   but couldn't.   someone is making a copy.   not sure who.

File: src/hilok.hpp (hunk 3 above)
Comment (earonesty) [reply src/hilok.hpp]: same as block=False, apparently: https://en.cppreference.com/w/cpp/thread/timed_mutex/try_lock_for

File: src/hilok.cpp (hunk 4 above)
Comment (earonesty) [reply src/hilok.cpp:103]: that would be "guard" like behavior.   which would be a different interface.   i handle the "with block" stuff at the python level, not the C++ level.   i suppose we could change the interface to be "with block" at C++ level too, but i wasn't sure how best to deal with that since __enter__ and __exit__ are different endpoints.   i think we need to leave it.

File: src/hilok.cpp (hunk 5 above)
Comment (earonesty) [reply src/hilok.cpp]: to get the reference counts down to 3.   but i really want them to be 2 .

File: src/hilok.cpp (hunk 6 above)
Comment (earonesty) [reply src/hilok.cpp:49]: first entry is "kref", increment happens after we save it

File: src/hilok.cpp (hunk 7 above)
Comment (earonesty) [reply src/hilok.cpp]: yeah, not needed, just checked.   was flailing trying to get the refcounts down to 2

File: tests/test.cpp (hunk 10 above)
Comment (earonesty) [reply tests/test.cpp:53]: done

File: tests/test.cpp (hunk 8 above)
Comment (earonesty) [reply tests/test.cpp:351]: done

File: src/hilok.cpp (hunk 7 above)
Comment (earonesty) [reply src/hilok.cpp]: removed

File: src/hilok.cpp (hunk 1 above)
Comment (earonesty) [reply src/hilok.cpp]: done

File: src/hilok.cpp (hunk 4 above)
Comment (earonesty) [reply src/hilok.cpp:128]: i think it can, if there's no situation where we want to be able to say:

h.write("x/y")

h.release_write("x/y")

which i think is not needed anywhere is it?

Comment (earonesty) [reply src/hilok.cpp:128]: but its super tricky.   the write() and read() functions would then have to be passed in the shared-pointer-to-manager as an argument - so they can construct a handle that holds a reference to the manager that they are a member of

Comment (earonesty) [reply src/hilok.cpp:128]: still didnt work, because python's with block doesn't hold the ref unless you explicitly hold it
